File: games/tictactoe/src/GameStateProvider/event.js
const createEventId = () => {
  let i = 0;

  return function createEventId() {
    return `${++i}`;
  };
};

export function subscribe(gameId) {
  return {
    id: createEventId(),
    payload: { Subscribe: gameId },
  };
}

export function unsubscribe(gameId) {
  return {
    id: createEventId(),
    payload: { Unsubscribe: gameId },
  };
}

export function get(gameId) {
  return {
    id: createEventId(),
    payload: { Get: gameId },
  };
}

export function set(gameId, gameState) {
  return {
    id: createEventId(),
    payload: { Set: [gameId, gameState] },
  };
}

File: server/src/ws/event.rs
use crate::game::{GameId, GameState};
use serde::{Deserialize, Serialize};

#[derive(Clone, Eq, PartialEq, Debug, Serialize, Deserialize, Default)]
#[serde(transparent)]
pub struct EventId(String);

#[derive(Clone, Serialize, Deserialize)]
pub struct Event<T> {
    pub id: EventId,
    pub payload: T,
}

#[derive(Clone, Serialize, Deserialize)]
pub enum Action {
    Subscribe(GameId),
    Unsubscribe(GameId),
    Get(GameId),
    Set(GameId, GameState),
}

#[derive(Clone, Serialize, Deserialize)]
pub enum Response {
    Update(GameState),
    Error(String), // TODO: error type? with code + context?
}
